src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/AzureManagedIdentityAuthenticationProvider.cs

```diff
@@ -0,0 +1,254 @@
+        }
+    }
+
+    #region IMDS Retry Helper
+    internal static class SqlManagedIdentityRetryHelper
+    {
+        internal const int DeltaBackOffInSeconds = 2;
+        internal const string RetryTimeoutError = "Reached retry timeout limit set by MsiRetryTimeout parameter in connection string.";
+
+        // for unit test purposes
+        internal static bool s_waitBeforeRetry = true;
+
+        internal static bool IsRetryableStatusCode(this HttpResponseMessage response)
+        {
+            // 404 NotFound, 429 TooManyRequests, and 5XX server error status codes are retryable
+            return Regex.IsMatch(((int)response.StatusCode).ToString(), @"404|429|5\d{2}");
+        }
+
+        /// <summary>
+        /// Implements recommended retry guidance here: https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/how-to-use-vm-token#retry-guidance
+        /// </summary>
+        internal static async Task<HttpResponseMessage> SendAsyncWithRetry(this HttpClient httpClient, Func<HttpRequestMessage> getRequest, int retryTimeoutInSeconds, int maxRetryCount, CancellationToken cancellationToken)
+        {
+            using (var timeoutTokenSource = new CancellationTokenSource())
+            using (var linkedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(timeoutTokenSource.Token, cancellationToken))
+            {
+                try
+                {
+                    // if retry timeout is configured, configure cancellation after timeout period elapses
+                    if (retryTimeoutInSeconds > 0)
+                    {
+                        timeoutTokenSource.CancelAfter(TimeSpan.FromSeconds(retryTimeoutInSeconds));
+                    }
+
+                    var attempts = 0;
+                    var backoffTimeInSecs = 0;
+                    HttpResponseMessage response;
+
+                    while (true)
+                    {
+                        attempts++;
+
+                        try
+                        {
+                            response = await httpClient.SendAsync(getRequest(), linkedTokenSource.Token).ConfigureAwait(false);
+
+                            if (response.IsSuccessStatusCode || !response.IsRetryableStatusCode() || attempts == maxRetryCount)
+                            {
+                                break;
+                            }
+                        }
+                        catch (HttpRequestException e)
+                        {
+                            if (attempts == maxRetryCount)
+                            {
+                                throw;
+                            }
+                            SqlClientEventSource.Log.TryTraceEvent("SendAsyncWithRetry | Exception occurred {0} | Attempting retry: {1} of {2}", e.Message, attempts, maxRetryCount);
+                        }
+
+                        if (s_waitBeforeRetry)
```

Comment:
Though this is for the unit test purpose, the following strategy is always executed. Is this necessary when we do retry? If it is only for testing, maybe we can enable this when in debug mode.

Reply:
I think you're talking about a different file (AADUtility), but I can answer it here.
I implemented retry in tests to avoid random errors since this retry policy is going to come into action when the token cannot be acquired in the first attempt. If we successfully acquire token in first attempt, no retry occurs, so this remains no-op, there are no side-effects.

Reply:
The code is duplicated in both files (AzureManagedIdentityAuthenticationProvider.cs and AADUtility.cs) which don't happen only in tests. The set of actions under the `s_waitBeforeRetry` condition will always be executed if the first attempt fails for some reason. It is acceptable if this retry logic is required all the time.